### DISCUSSION TITLE
[BUG] web build fails with 'Named export 'APIKeyApi' not found'

### DISCUSSION THREAD

**Discussion by waclaw66** [immich-app/immich#6396] [Q&A] [answered]
### The bug

I'm trying to build web according `immich/server/Dockerfile` as usual, however it fails on `npm run build` with following message...
Not sure what could be wrong, open-api is already built.

```
node:internal/event_target:1088
  process.nextTick(() => { throw err; });
                           ^
file:///home/waclaw/src/immich/web/.svelte-kit/output/server/chunks/api.js:1
import { configuration, ActivityApi, AlbumApi, AuditApi, LibraryApi, AssetApi, AuthenticationApi, FaceApi, JobApi, APIKeyApi, OAuthApi, PartnerApi, SearchApi, ServerInfoApi, SharedLinkApi, PersonApi, SystemConfigApi, UserApi, common, base, JobName, AssetJobName } from "@immich/sdk";
                                                                                                                   ^^^^^^^^^
SyntaxError: Named export 'APIKeyApi' not found. The requested module '@immich/sdk' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@immich/sdk';
const { configuration, ActivityApi, AlbumApi, AuditApi, LibraryApi, AssetApi, AuthenticationApi, FaceApi, JobApi, APIKeyApi, OAuthApi, PartnerApi, SearchApi, ServerInfoApi, SharedLinkApi, PersonApi, SystemConfigApi, UserApi, common, base, JobName, AssetJobName } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:132:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:214:5)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async analyse (file:///home/waclaw/src/immich/web/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js:60:16)
    at async MessagePort.<anonymous> (file:///home/waclaw/src/immich/web/node_modules/@sveltejs/kit/src/utils/fork.js:22:16)
Emitted 'error' event on Worker instance at:
    at [kOnErrorMessage] (node:internal/worker:326:10)
    at [kOnMessage] (node:internal/worker:337:37)
    at MessagePort.<anonymous> (node:internal/worker:232:57)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:814:20)
    at exports.emitMessage (node:internal/per_context/messageport:23:28)
```

When I try to run `npm run dev`, it fails with...
```
Unexpected token 'export'

/home/waclaw/src/immich/web/node_modules/@immich/sdk/build/esm/index.js:1
export * from './client';
^^^^^^

SyntaxError: Unexpected token 'export'
    at internalCompileFunction (node:internal/vm:77:18)
    at wrapSafe (node:internal/modules/cjs/loader:1288:20)
    at Module._compile (node:internal/modules/cjs/loader:1340:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at cjsLoader (node:internal/modules/esm/translators:345:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:294:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24
```

### The OS that Immich Server is running on

Fedora 39

### Version of Immich Server

main

### Version of Immich Mobile App

not relevant

### Platform with the issue

- [ ] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
not relevant
```

### Your .env content

```Shell
not relevant
```

### Reproduction steps

```bash
manual execution of commands from immich/server/Dockerfile
```

**Comment by martabal** [comment]
Have you moved any folders ? Just updated the AIO image for v1.93.0 without any issue

**Comment by waclaw66** [comment]
Not sure, what folders do you mean? I haven't moved any folders, I'm just trying to execute build commands from `server/Dockerfile` without using Docker.

**Comment by martabal** [comment]
> Not sure, what folders do you mean?

Root folders in the repo such as `web`, `server`, `open-api` ...

> I'm just trying to execute build commands from server/Dockerfile

Oh sorry, according to your logs, I thought the error occurred when building the web app

**Comment by waclaw66** [comment]
> Root folders in the repo such as `web`, `server`, `open-api` ...

Haven't touch them.
 
> Oh sorry, according to your logs, I thought the error occurred when building the web app

Yes, I'm building web app, open-api and server has been built without any error, but web is failing.

**Comment by jrasm91** [comment]
How are you building open API? And what command are you using to build the server image?

**Reply by jrasm91** [reply]
Also, it should not be importing the sdk as a common js module, but an esm. If it is trying to do it via a common js module it may not be building correctly or you are using an old version of nodejs or typescript.

**Reply by waclaw66** [reply]
Commands from `server/Dockerfile`
```
git clone https://github.com/immich-app/immich.git

cd immich/server
npm ci
npm install --no-save @img/sharp-linux-x64
npm run build

cd immich/open-api/typescript-sdk
npm ci
npm run build

cd immich/web
npm ci
npm run build
```
using Fedora 39 disto Node v20.10.0 and typescript installed by package.json deps.
It worked fine until #6334.
I'm aware that I'm not using suggested installation method, however it supposed to work when I'm reproducing docker commands as well.

**Reply by jrasm91** [reply]
Using those same commands on Ubuntu works fine for me.

**Reply by waclaw66** [reply]
What version of node.js do you use?

**Reply by jrasm91** [reply]
I tested it with 20.9 and 20.10.

**Reply by jrasm91** [reply]
The problem looks like it is caused by the typescript-sdk being imported to the web package as a commonjs module, not an esm. The original error you reported is in `.svelte-kit/output/server/chunks/api.js:1` and line one looks like this for me:

```typescript
import globalAxios from "axios";
var __extends$1 = globalThis && globalThis.__extends || function() {
  var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function(d2, b2) {
      d2.__proto__ = b2;
```

This is because `typescript-sdk` has a way to import it as an esm module and the `moduleResolution` is set to `Bundler`. The source files get pulled into the bundle. In your error, you have this line:

```
import { configuration, ActivityApi, AlbumApi, AuditApi, LibraryApi, AssetApi, AuthenticationApi, FaceApi, JobApi, APIKeyApi, OAuthApi, PartnerApi, SearchApi, ServerInfoApi, SharedLinkApi, PersonApi, SystemConfigApi, UserApi, common, base, JobName, AssetJobName } from "@immich/sdk";
                                                                                                                   ^^^^^^^^^
SyntaxError: Named export 'APIKeyApi' not found. The requested module '@immich/sdk' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:
```

Which is strange, because this means it is trying to import it as a commonjs module.

**Reply by waclaw66** [reply]
That's strange, you should get same result using the same nodejs version.

**Reply by jrasm91** [reply]
Yeah, I have no idea why it is doing something different on your system. I assume you have no other uncommitted changes, right? And no other customizations in your build?

**Reply by waclaw66** [reply, marked as answer]
It turned out I had installed old 9.1.3 version of `npm` in `/usr/local/bin` instead using system wide 10.2.3. Everything works fine with that version.
Thanks for directing me!
